util/reflect.go

```diff
@@ -367,61 +413,368 @@ type NodeInfo struct {
+	case IsTypeStructPtr(t):
+		t = t.Elem()
+		if !IsNilOrInvalidValue(v) {
+			v = v.Elem()
+		}
+		fallthrough
+	case IsTypeStruct(t):
+		for i := 0; i < t.NumField(); i++ {
+			sf := t.Field(i)
+			nn := &NodeInfo{
+				Parent:      ni,
+				StructField: sf,
+				FieldValue:  reflect.Zero(sf.Type),
+			}
+			if !IsNilOrInvalidValue(v) {
+				nn.FieldValue = v.Field(i)
+			}
+			ps, err := SchemaPaths(nn.StructField)
+			if err != nil {
+				return NewErrs(err)
+			}
+			for _, p := range ps {
+				nn.Schema = ChildSchema(ni.Schema, p)
+				if nn.Schema == nil {
+					e := fmt.Errorf("forEachFieldInternal could not find child schema with path %v from schema name %s", p, ni.Schema.Name)
+					DbgPrint(e.Error())
+					return NewErrs(e)
+				}
+				nn.PathFromParent = p
+				// In the case of a map/slice, the path is of the form
+				// "container/element".
+				if IsTypeSlice(sf.Type) || IsTypeMap(sf.Type) {
+					nn.PathFromParent = p[0:1]
+				}
+				errs = AppendErrs(errs, forEachFieldInternal(nn, in, out, iterFunction))
+			}
 		}
 
-	case IsValueSlice(ni.FieldValue):
-		for i := 0; i < ni.FieldValue.Len(); i++ {
-			nn := *ni
-			nn.FieldValue = ni.FieldValue.Index(i)
-			errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+	case IsTypeSlice(t):
+		schema := *(ni.Schema)
+		schema.ListAttr = nil
+		var pp []string
+		// Leaf-list elements share the parent schema.
+		if !schema.IsLeafList() {
+			pp = []string{schema.Name}
+		}
+		if IsNilOrInvalidValue(v) {
+			// Traverse the type tree only from this point.
+			nn := &NodeInfo{
+				Parent:         ni,
+				PathFromParent: pp,
+				Schema:         &schema,
+				FieldValue:     reflect.Zero(t.Elem()),
+			}
+			errs = AppendErrs(errs, forEachFieldInternal(nn, in, out, iterFunction))
+		} else {
+			for i := 0; i < ni.FieldValue.Len(); i++ {
+				nn := *ni
+				// The schema for each element is the list schema minus the list
+				// attrs.
+				nn.Schema = &schema
+				nn.Parent = ni
+				nn.PathFromParent = pp
+				nn.FieldValue = ni.FieldValue.Index(i)
+				errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+			}
 		}
 
-	case IsValueMap(ni.FieldValue):
-		for _, key := range ni.FieldValue.MapKeys() {
-			nn := *ni
-			nn.FieldValue = ni.FieldValue.MapIndex(key)
-			nn.FieldKey = key
-			nn.FieldKeys = ni.FieldValue.MapKeys()
-			errs = AppendErrs(errs, forEachFieldInternal(&nn, in, out, iterFunction))
+	case IsTypeMap(t):
+		schema := *(ni.Schema)
+		schema.ListAttr = nil
+		if IsNilOrInvalidValue(v) {
+			nn := &NodeInfo{
+				Parent:         ni,
+				PathFromParent: []string{schema.Name},
+				Schema:         &schema,
+				FieldValue:     reflect.Zero(t.Elem()),
+			}
+			errs = AppendErrs(errs, forEachFieldInternal(nn, in, out, iterFunction))
+		} else {
+			for _, key := range ni.FieldValue.MapKeys() {
+				nn := *ni
+				nn.Schema = &schema
+				nn.Parent = ni
+				nn.PathFromParent = []string{schema.Name}
```

Comment:
Consider making the `PathFromParent` more explicit. Is there a reason that it needs to be `[]string` rather than `[]*gnmipb.PathElem`, since this would allow `FieldKey` to be specified within the actual path specified.